File: configuration/src/main/java/com/synopsys/integration/configuration/property/base/ValuedProperty.java
package com.synopsys.integration.configuration.property.base;

import org.jetbrains.annotations.NotNull;

import com.synopsys.integration.configuration.parse.ValueParser;

/**
 * A property that returns null when it is not present in a Configuration.
 * @param <R> the type this property returns when it is retrieved from a Configuration.
 * @param <V> the type this property uses when parsing user input.
 */
public abstract class ValuedProperty<V, R> extends TypedProperty<V, R> {
    private final V defaultValue;

    public ValuedProperty(@NotNull String key, @NotNull ValueParser<V> valueParser, V defaultValue) {
        super(key, valueParser);
        this.defaultValue = defaultValue;
    }

    public V getDefaultValue() {
        return defaultValue;
    }

}

File: configuration/src/main/java/com/synopsys/integration/configuration/property/base/ValuedProperty.kt
package com.synopsys.integration.configuration.property.base

import com.synopsys.integration.configuration.parse.ValueParser

abstract class ValuedProperty<T>(key: String, parser: ValueParser<T>, val default: T) : TypedProperty<T>(key, parser) {
    // This is a property with a key and with a default value, it will always have a value.

    var replacesDeprecatedProperty: ValuedProperty<T>? = null
    fun replacesDeprecatedProperty(oldProperty: ValuedProperty<T>): ValuedProperty<T> {
        this.replacesDeprecatedProperty = oldProperty
        return this
    }
}